internal/pkg/cli/deploy/workload.go

```diff
@@ -374,22 +388,85 @@ func (d *workloadDeployer) uploadContainerImages(out *UploadArtifactsOutput) err
 	if err != nil {
 		return fmt.Errorf("login to image repository: %w", err)
 	}
+
+	var mu sync.Mutex
 	out.ImageDigests = make(map[string]ContainerImageIdentifier, len(buildArgsPerContainer))
+
+	// counter for indexing the labeled output buffers.
+	count := 0
+	// An array of buffers, one for each container image build and push ouput.
+	labeledBuffers := make([]*syncbuffer.LabeledSyncBuffer, len(buildArgsPerContainer))
+
+	// create a context and an error group.
+	g, ctx := errgroup.WithContext(context.Background())
+
+	// create a new cursor and hide it.
+	cursor := cursor.New()
+	cursor.Hide()
+
 	for name, buildArgs := range buildArgsPerContainer {
+
+		// create a copy of loop variables to avoid data race.
+		name := name
+		buildArgs := buildArgs
+
+		// create a pipe for streaming the build and push output.
+		pr, pw := io.Pipe()
+
+		// generate build args slice for docker build label.
 		buildArgsList, err := buildArgs.GenerateDockerBuildArgs(dockerengine.New(exec.NewCmd()))
 		if err != nil {
 			return fmt.Errorf("generate docker build args: %w", err)
 		}
-		log.Infof(dockerengine.DockerBuildLabel(buildArgs.Platform, buildArgsList))
-		digest, err := d.repository.BuildAndPush(buildArgs)
-		if err != nil {
-			return fmt.Errorf("build and push image: %w", err)
-		}
-		out.ImageDigests[name] = ContainerImageIdentifier{
-			Digest:            digest,
-			CustomTag:         d.image.CustomTag,
-			GitShortCommitTag: d.image.GitShortCommitTag,
+		buf := syncbuffer.New()
+		labeledBuffer := buf.WithLabel(fmt.Sprintf("Building your container image: docker %s\n", strings.Join(buildArgsList, " ")))
+		labeledBuffers[count] = labeledBuffer
+		count++
+
+		g.Go(func() error {
+			defer pw.Close()
+			digest, err := d.repository.BuildAndPush(ctx, buildArgs, pw)
+			if err != nil {
+				return fmt.Errorf("build and push image: %w", err)
+			}
+			mu.Lock()
+			out.ImageDigests[name] = ContainerImageIdentifier{
+				Digest:            digest,
+				CustomTag:         d.image.CustomTag,
+				GitShortCommitTag: d.image.GitShortCommitTag,
+			}
+			mu.Unlock()
+			return nil
+		})
+		g.Go(func() error {
+			if err := buf.Copy(pr); err != nil {
+				return fmt.Errorf("copying build and push output for container %s: %w", name, err)
```

Comment:
I like that we show which container we are talking about ❤️ . We should maximize this practice in this case so that in every step customers know what container logs they are looking at.